src/t8_forest/t8_forest_vtk.cxx

```diff
@@ -94,6 +122,190 @@ typedef int         (*t8_forest_vtk_cell_data_kernel) (t8_forest_t forest,
                                                        void **data,
                                                        T8_VTK_KERNEL_MODUS
                                                        modus);
+#if T8_WITH_VTK
+void
+t8_write_vtk_via_API (t8_forest_t forest, const char *fileprefix)
+{
+  /*Check assertions: forest and fileprefix are not NULL and forest is commited */
+  T8_ASSERT (forest != NULL);
+  T8_ASSERT (forest->rc.refcount > 0);
+  T8_ASSERT (forest->committed);
+  T8_ASSERT (fileprefix != NULL);
+
+  long int            point_id = 0;     /* The id of the point in the points Object. */
+  t8_cmesh_t          cmesh;
+  t8_locidx_t         ielement;
+  t8_tree_t           tree;
+  t8_locidx_t         itree, ivertex;
+  t8_ctree_t          ctree;
+  double             *vertices;
+  double              coordinates[3];
+  double              x, y, z;
+  int                 elem_id = 0;
+
+/* Since we want to use different Element types and a points and cellArray 
+ * we have to declare them. The cellArray stores the Elements.*/
+  vtkNew < vtkPoints > points;
+  vtkNew < vtkCellArray > cellArray;
+  vtkNew < vtkHexahedron > hexa;
+  vtkNew < vtkVertex > vertex;
+  vtkNew < vtkLine > line;
+  vtkNew < vtkQuad > quad;
+  vtkNew < vtkTriangle > tri;
+  vtkNew < vtkPyramid > pyramid;
+  vtkNew < vtkWedge > prism;
+  vtkNew < vtkTetra > tet;
+
+  std::string filename = fileprefix;
+  // Append process number
+  filename += "_" + std::to_string (forest->mpirank);
+
+  t8_debugf ("Entering vtk API output with filename %s\n", fileprefix);
+
+  filename += ".vtu";
+  cmesh = t8_forest_get_cmesh (forest);
+  /* The cellTypes Array stores the element types as integers(see vtk doc) */
+  int                *cellTypes =
+    T8_ALLOC (int, t8_forest_get_num_element (forest));
+/* We iterate over all local trees*/
+  for (itree = 0;
+       (int) itree < t8_forest_get_num_local_trees (forest); itree++) {
+
+    vertices = t8_forest_get_tree_vertices (forest, itree);
+    tree = t8_forest_get_tree (forest, itree);
+    t8_eclass_scheme_c *scheme =
+      t8_forest_get_eclass_scheme (forest, tree->eclass);
+    t8_locidx_t         elems_in_tree =
+      (t8_locidx_t) t8_element_array_get_count (&tree->elements);
+    /* We iterate over all elements in the tree */
+    for (ielement = 0; ielement < elems_in_tree; ielement++) {
+      t8_element_t       *element =
+        t8_forest_get_element_in_tree (forest, itree, ielement);
+      T8_ASSERT (element != NULL);
+      t8_element_shape_t  element_shape = scheme->t8_element_shape (element);
+      /* For each element we iterate over all points */
+      for (ivertex = 0; ivertex < t8_eclass_num_vertices[tree->eclass];
+           ivertex++, point_id++) {
```

Comment:
Use `t8_element_num_corners` instead of `t8_eclass_num_vertices`.
The eclass is only for the tree.
Imagine a pyramid tree that has a tet element.
You want 4 vertices for the tet. The eclass will however give you 5 since it stores 'pyramid'.

Reply:
I get '‘t8_element_num_corners’ was not declared in this scope' even though we have included:
#include <t8_schemes/t8_default_cxx.hxx>
#include <t8_schemes/t8_default/t8_default_common_cxx.hxx>

Reply:
This function is part of the element scheme, so you need to call it as `scheme->t8_element_num_corners`. This should work.